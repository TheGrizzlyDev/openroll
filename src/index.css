@tailwind base;
@tailwind components;
@tailwind utilities;

body {
  background: var(--color-bg);
  color: var(--color-text);
  font-family: var(--font-body);
  margin: 0;
}

* {
  box-sizing: border-box;
}

@media (width <= 400px) {
  body {
    font-size: 14px;
  }

  body h1 {
    font-size: 1.5rem;
  }

  body h2 {
    font-size: 1.25rem;
  }

  body .tabs button {
    font-size: 0.875rem;
    padding: 0.25rem;
  }

  body .dice-roller {
    align-items: stretch;
    flex-direction: column;
    gap: 0.25rem;
  }

  body .base-input,
  body button {
    font-size: 0.875rem;
    padding: 0.25rem;
  }

  body .inventory-form {
    flex-direction: column;
  }

  body .overlay {
    font-size: 0.875rem;
    padding: 0.4rem 0.6rem;
  }
}
