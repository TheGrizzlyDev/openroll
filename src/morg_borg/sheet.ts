export const createSheet = () => ({
  name: '',
  class: '',
  str: 0,
  agi: 0,
  pre: 0,
  tou: 0,
  statDice: {
    str: '1d20',
    agi: '1d20',
    pre: '1d20',
    tou: '1d20'
  },
  hp: 0,
  maxHp: 0,
  armor: 0,
  omens: 0,
  silver: 0,
  notes: ''
})
