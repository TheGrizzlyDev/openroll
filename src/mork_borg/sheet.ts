export interface Sheet {
  name: string
  class: string
  str: number
  agi: number
  pre: number
  tou: number
  hp: number
  tempHp: number
  maxHp: number
  armor: number
  omens: number
  silver: number
  trait: string
  background: string
  notes: string
  conditions: string[]
}

export const createSheet = (): Sheet => ({
  name: '',
  class: '',
  str: 0,
  agi: 0,
  pre: 0,
  tou: 0,
  hp: 0,
  tempHp: 0,
  maxHp: 0,
  armor: 0,
  omens: 0,
  silver: 0,
  trait: '',
  background: '',
  notes: '',
  conditions: []
})
