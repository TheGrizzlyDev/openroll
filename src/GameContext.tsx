/* eslint react-refresh/only-export-components: off */
import {
  createContext,
  useContext,
  useCallback,
  useEffect,
  useRef,
  useReducer,
  type ReactNode,
  type Dispatch
} from 'react'
import { DiceRoller as DiceParser, type DiceRoll } from '@dice-roller/rpg-dice-roller'
import { useNavigate, useLocation } from 'react-router-dom'
import { createSheet, type Sheet } from './morg_borg/sheet'
import { generateCharacter } from './morg_borg/generateCharacter'
import type { MutableRefObject } from 'react'

export interface InventoryItem {
  id: number
  name: string
  qty: number
  notes: string
}

export interface Scroll {
  id: number
  type: 'unclean' | 'sacred'
  name: string
  casts: number
  notes: string
}

export interface LogEntry {
  label: string
  notation?: string
  output: string
  total?: number
}

export interface Character {
  name: string
  sheet: Sheet
  inventory: InventoryItem[]
  scrolls: Scroll[]
}

export interface OverlayState {
  message: string
  visible: boolean
}

export interface GameState {
  characters: Character[]
  current: number | null
  sheet: Sheet
  inventory: InventoryItem[]
  scrolls: Scroll[]
  log: LogEntry[]
  activeTab: string
  overlay: OverlayState
}

export type GameAction =
  | { type: 'SET_CHARACTERS'; characters: Character[] }
  | { type: 'SET_CURRENT'; current: number | null }
  | { type: 'SET_SHEET'; sheet: Sheet }
  | { type: 'SET_INVENTORY'; inventory: InventoryItem[] }
  | { type: 'SET_SCROLLS'; scrolls: Scroll[] }
  | { type: 'ADD_LOG'; entry: LogEntry }
  | { type: 'SET_LOG'; log: LogEntry[] }
  | { type: 'SET_ACTIVE_TAB'; tab: string }
  | { type: 'SET_OVERLAY'; overlay: OverlayState }

export interface GameContextValue {
  state: GameState
  dispatch: Dispatch<GameAction>
  overlayTimeout: MutableRefObject<ReturnType<typeof setTimeout> | null>
  loadCharacter: (_idx: number, _opts?: { navigate?: boolean }) => void
  createCharacter: (_class?: string) => void
  finalizeCharacter: () => void
  cancelCreation: () => void
  deleteCharacter: (_idx: number) => void
  exportCharacters: () => string
  importCharacters: (_data: unknown) => boolean
  roll: (_notation: string, _label?: string) => number
  logInventory: (_message: string) => void
}

export const GameContext = createContext<GameContextValue | null>(null)
export const useGameContext = () => useContext(GameContext) as GameContextValue

export function GameProvider({ children }: { children: ReactNode }) {
  const initialState: GameState = {
    characters: (() => {
      const saved = localStorage.getItem('characters')
      return saved ? JSON.parse(saved) : []
    })(),
    current: null,
    sheet: createSheet(),
    inventory: [],
    scrolls: [],
    log: (() => {
      const saved = localStorage.getItem('log')
      return saved ? JSON.parse(saved) : []
    })(),
    activeTab: 'character',
    overlay: { message: '', visible: false }
  }

  const reducer = (state: GameState, action: GameAction): GameState => {
    switch (action.type) {
      case 'SET_CHARACTERS':
        return { ...state, characters: action.characters }
      case 'SET_CURRENT':
        return { ...state, current: action.current }
      case 'SET_SHEET':
        return { ...state, sheet: action.sheet }
      case 'SET_INVENTORY':
        return { ...state, inventory: action.inventory }
      case 'SET_SCROLLS':
        return { ...state, scrolls: action.scrolls }
      case 'ADD_LOG':
        return { ...state, log: [action.entry, ...state.log] }
      case 'SET_LOG':
        return { ...state, log: action.log }
      case 'SET_ACTIVE_TAB':
        return { ...state, activeTab: action.tab }
      case 'SET_OVERLAY':
        return { ...state, overlay: action.overlay }
      default:
        return state
    }
  }

  const [state, dispatch] = useReducer(reducer, initialState)
  const overlayTimeout = useRef<ReturnType<typeof setTimeout> | null>(null)

  const skipSave = useRef<boolean>(false)
  const charactersRef = useRef<Character[]>(state.characters)

  useEffect(() => {
    charactersRef.current = state.characters
  }, [state.characters])

  useEffect(() => {
    if (skipSave.current) return
    localStorage.setItem('characters', JSON.stringify(state.characters))
  }, [state.characters])

  useEffect(() => {
    if (skipSave.current) return
    localStorage.setItem('log', JSON.stringify(state.log))
  }, [state.log])

  const { sheet, inventory, scrolls, current } = state

  useEffect(() => {
    if (current === null) return
    if (skipSave.current) return
    const updated = [...charactersRef.current]
    const existing =
      updated[current] || { name: '', sheet: createSheet(), inventory: [], scrolls: [] }
    updated[current] = {
      ...existing,
      name: sheet.name,
      sheet,
      inventory,
      scrolls
    }
    dispatch({ type: 'SET_CHARACTERS', characters: updated })
  }, [sheet, inventory, scrolls, current])

  const navigate = useNavigate()
  const location = useLocation()

  const loadCharacter = useCallback(
    (idx: number, { navigate: doNavigate = true }: { navigate?: boolean } = {}) => {
      const char = charactersRef.current[idx]
      if (!char) return
      skipSave.current = true
      dispatch({ type: 'SET_CURRENT', current: idx })
      dispatch({ type: 'SET_SHEET', sheet: char.sheet || createSheet() })
      dispatch({ type: 'SET_INVENTORY', inventory: char.inventory || [] })
      dispatch({ type: 'SET_SCROLLS', scrolls: char.scrolls || [] })
      if (doNavigate) navigate(`/sheet/${idx}`)
      setTimeout(() => {
        skipSave.current = false
      })
    },
    [navigate]
  )

  const createCharacter = (cls?: string) => {
    skipSave.current = true
    const {
      sheet: newSheet,
      inventory: newInv,
      scrolls: newScrolls
    } = generateCharacter(cls)
    const index = state.characters.length
    dispatch({ type: 'SET_SHEET', sheet: newSheet })
    dispatch({ type: 'SET_INVENTORY', inventory: newInv })
    dispatch({ type: 'SET_SCROLLS', scrolls: newScrolls })
    dispatch({ type: 'SET_CURRENT', current: index })
    navigate('/generator')
  }

  const finalizeCharacter = () => {
    const index = state.current ?? state.characters.length
    const updated = [...state.characters]
    updated[index] = {
      name: '',
      sheet: state.sheet,
      inventory: state.inventory,
      scrolls: state.scrolls
    }
    dispatch({ type: 'SET_CHARACTERS', characters: updated })
    skipSave.current = false
    navigate(`/sheet/${index}`)
  }

  const cancelCreation = () => {
    skipSave.current = false
    dispatch({ type: 'SET_CURRENT', current: null })
    navigate('/characters')
  }

  const deleteCharacter = (idx: number) => {
    const updated = state.characters.filter((_, i) => i !== idx)
    dispatch({ type: 'SET_CHARACTERS', characters: updated })
    dispatch({ type: 'SET_CURRENT', current: null })
    navigate('/characters')
  }

  const exportCharacters = () => JSON.stringify(state.characters, null, 2)

  const importCharacters = (data: unknown) => {
    let parsed: unknown
    try {
      parsed = typeof data === 'string' ? JSON.parse(data) : data
    } catch {
      return false
    }
    if (!Array.isArray(parsed)) return false
    const sanitized = parsed.map(c => {
      const name = typeof (c as Character).name === 'string' ? (c as Character).name : ''
      const sheet =
        typeof (c as Character).sheet === 'object'
          ? { ...createSheet(), ...(c as Character).sheet }
          : createSheet()
      const inventory = Array.isArray((c as Character).inventory)
        ? (c as Character).inventory
        : []
      const scrolls = Array.isArray((c as Character).scrolls)
        ? (c as Character).scrolls
        : []
      return { name, sheet, inventory, scrolls }
    })
    dispatch({ type: 'SET_CHARACTERS', characters: [...state.characters, ...sanitized] })
    return true
  }

  const roller = new DiceParser()

  const roll = (notation: string, label = '') => {
    const result = roller.roll(notation) as DiceRoll
    const entry: LogEntry = { label, notation, output: result.output, total: result.total }
    dispatch({ type: 'ADD_LOG', entry })
    const message = `${label ? `${label}: ` : ''}${result.output}`
    dispatch({ type: 'SET_OVERLAY', overlay: { message, visible: true } })
    if (overlayTimeout.current) clearTimeout(overlayTimeout.current)
    overlayTimeout.current = setTimeout(
      () => dispatch({ type: 'SET_OVERLAY', overlay: { ...state.overlay, visible: false } }),
      10000
    )
    return result.total as number
  }

  const logInventory = (message: string) => {
    dispatch({ type: 'ADD_LOG', entry: { label: 'Inventory', output: message } })
  }

  useEffect(() => {
    if (location.pathname === '/characters') {
      dispatch({ type: 'SET_CURRENT', current: null })
    }
  }, [location.pathname])

  const value = {
    state,
    dispatch,
    overlayTimeout,
    loadCharacter,
    createCharacter,
    finalizeCharacter,
    cancelCreation,
    deleteCharacter,
    exportCharacters,
    importCharacters,
    roll,
    logInventory
  }

  return (
    <GameContext.Provider value={value}>
      {children}
    </GameContext.Provider>
  )
}
