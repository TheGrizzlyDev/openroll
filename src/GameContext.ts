import { create, type StateCreator } from 'zustand'
import { devtools, persist, createJSONStorage, type NamedSet } from 'zustand/middleware'
import { DiceRoller as DiceParser, type DiceRoll } from '@dice-roller/rpg-dice-roller'
import { createSheet, type Sheet } from './mork_borg/sheet'
import { generateCharacter } from './mork_borg/generateCharacter'
import type { ApplyNode } from './oml/parser'

export interface InventoryItem {
  id: number
  name: string
  qty: number
  notes: string
}

export interface Scroll {
  id: number
  type: 'unclean' | 'sacred'
  name: string
  casts: number
  notes: string
}

export interface LogEntry {
  label: string
  notation?: string
  output: string
  total?: number
}

export interface Character {
  id: string
  name: string
  sheet: Sheet
  inventory: InventoryItem[]
  scrolls: Scroll[]
}

export type DiceType = 'd4' | 'd6' | 'd8' | 'd12' | 'd20'

export interface OverlayState {
  message: string
  roll: { dice: Array<{ type: DiceType; result: number }>; total: number } | null
  visible: boolean
}

export interface DiceStyle {
  color: string
  edgeColor: string
  textureUrls: string[]
}

export interface GameState {
  characters: Character[]
  lastAccess: Record<string, number>
  current: number | null
  sheet: Sheet
  inventory: InventoryItem[]
  scrolls: Scroll[]
  log: LogEntry[]
  activeTab: string
  overlay: OverlayState
  diceStyle: DiceStyle
}

export type GameAction =
  | { type: 'SET_CHARACTERS'; characters: Character[] }
  | { type: 'SET_CURRENT'; current: number | null }
  | { type: 'SET_SHEET'; sheet: Sheet }
  | { type: 'SET_INVENTORY'; inventory: InventoryItem[] }
  | { type: 'SET_SCROLLS'; scrolls: Scroll[] }
  | { type: 'ADD_LOG'; entry: LogEntry }
  | { type: 'SET_LOG'; log: LogEntry[] }
  | { type: 'SET_ACTIVE_TAB'; tab: string }
  | { type: 'SET_OVERLAY'; overlay: OverlayState }
  | { type: 'SET_DICE_STYLE'; diceStyle: DiceStyle }

export interface GameContextValue {
  state: GameState
  dispatch: (_action: GameAction) => void
  overlayTimeout: ReturnType<typeof setTimeout> | null
  setOverlayTimeout: (_t: ReturnType<typeof setTimeout> | null) => void
  loadCharacter: (_idx: number) => void
  createCharacter: (_class?: string) => void
  finalizeCharacter: () => number
  cancelCreation: () => void
  deleteCharacter: (_idx: number) => void
  exportCharacters: () => string
  importCharacters: (_data: unknown) => boolean
  roll: (_notation: string, _label?: string) => { total: number; output: string }
  logInventory: (_message: string) => void
  applyEffect: (_effect: ApplyNode) => number
  setDiceStyle: (_style: DiceStyle) => void
}

const initialState: GameState = {
  characters: [],
  lastAccess: {},
  current: null,
  sheet: createSheet(),
  inventory: [],
  scrolls: [],
  log: [],
  activeTab: 'character',
  overlay: { message: '', roll: null, visible: false },
  diceStyle: { color: '#ffffff', edgeColor: '#000000', textureUrls: [] }
}

const reducer = (state: GameState, action: GameAction): GameState => {
  switch (action.type) {
    case 'SET_CHARACTERS':
      return { ...state, characters: action.characters }
    case 'SET_CURRENT':
      return { ...state, current: action.current }
    case 'SET_SHEET': {
      const newState = { ...state, sheet: action.sheet }
      if (state.current !== null) {
        const updated = [...state.characters]
        let char = updated[state.current]
        if (!char) {
          char = {
            id: crypto.randomUUID(),
            name: '',
            sheet: createSheet(),
            inventory: [],
            scrolls: []
          }
        } else if (!char.id) {
          char = { ...char, id: crypto.randomUUID() }
        }
        char = {
          ...char,
          name: action.sheet.name,
          sheet: action.sheet,
          inventory: state.inventory,
          scrolls: state.scrolls
        }
        updated[state.current] = char
        const lastAccess = { ...state.lastAccess, [char.id]: Date.now() }
        newState.characters = updated
        newState.lastAccess = lastAccess
      }
      return newState
    }
    case 'SET_INVENTORY': {
      const newState = { ...state, inventory: action.inventory }
      if (state.current !== null) {
        const updated = [...state.characters]
        let char = updated[state.current]
        if (!char) {
          char = {
            id: crypto.randomUUID(),
            name: '',
            sheet: createSheet(),
            inventory: [],
            scrolls: []
          }
        } else if (!char.id) {
          char = { ...char, id: crypto.randomUUID() }
        }
        char = { ...char, inventory: action.inventory }
        updated[state.current] = char
        const lastAccess = { ...state.lastAccess, [char.id]: Date.now() }
        newState.characters = updated
        newState.lastAccess = lastAccess
      }
      return newState
    }
    case 'SET_SCROLLS': {
      const newState = { ...state, scrolls: action.scrolls }
      if (state.current !== null) {
        const updated = [...state.characters]
        let char = updated[state.current]
        if (!char) {
          char = {
            id: crypto.randomUUID(),
            name: '',
            sheet: createSheet(),
            inventory: state.inventory,
            scrolls: []
          }
        } else if (!char.id) {
          char = { ...char, id: crypto.randomUUID() }
        }
        char = { ...char, scrolls: action.scrolls }
        updated[state.current] = char
        const lastAccess = { ...state.lastAccess, [char.id]: Date.now() }
        newState.characters = updated
        newState.lastAccess = lastAccess
      }
      return newState
    }
    case 'ADD_LOG':
      return { ...state, log: [action.entry, ...state.log] }
    case 'SET_LOG':
      return { ...state, log: action.log }
    case 'SET_ACTIVE_TAB':
      return { ...state, activeTab: action.tab }
    case 'SET_OVERLAY':
      return { ...state, overlay: action.overlay }
    case 'SET_DICE_STYLE':
      return { ...state, diceStyle: action.diceStyle }
    default:
      return state
  }
}

const storeCreator: StateCreator<
  GameContextValue,
  [],
  [["zustand/devtools", never], ["zustand/persist", unknown]]
> = (set: NamedSet<GameContextValue>, get) => ({
  state: initialState,
  dispatch: (action: GameAction) =>
    set(state => ({ state: reducer(state.state, action) }), false, action.type),
  overlayTimeout: null,
  setOverlayTimeout: t => set({ overlayTimeout: t }, false, 'setOverlayTimeout'),
  loadCharacter: idx => {
    const char = get().state.characters[idx]
    if (!char) return
    const now = Date.now()
    const id = char.id ?? crypto.randomUUID()
    set(({ state }) => {
      const characters = [...state.characters]
      if (!char.id) {
        characters[idx] = { ...char, id }
      }
      const lastAccess = { ...state.lastAccess, [id]: now }
      return {
        state: {
          ...state,
          characters,
          lastAccess,
          current: idx,
          sheet: char.sheet || createSheet(),
          inventory: char.inventory || [],
          scrolls: char.scrolls || []
        }
      }
    }, false, 'loadCharacter')
  },
  createCharacter: cls => {
    const { sheet, inventory, scrolls } = generateCharacter(cls)
    const index = get().state.characters.length
    set(({ state }) => ({
      state: {
        ...state,
        sheet,
        inventory,
        scrolls,
        current: index
      }
    }), false, 'createCharacter')
  },
  finalizeCharacter: () => {
    const { state } = get()
    const index = state.current ?? state.characters.length
    const updated = [...state.characters]
    const existing = updated[index]
    const id = existing?.id ?? crypto.randomUUID()
    updated[index] = {
      id,
      name: '',
      sheet: state.sheet,
      inventory: state.inventory,
      scrolls: state.scrolls
    }
    const lastAccess = { ...state.lastAccess, [id]: Date.now() }
    set(
      {
        state: { ...state, characters: updated, lastAccess, current: index }
      },
      false,
      'finalizeCharacter'
    )
    return index
  },
  cancelCreation: () =>
    set(({ state }) => ({ state: { ...state, current: null } }), false, 'cancelCreation'),
  deleteCharacter: idx =>
    set(({ state }) => {
      const characters = state.characters.filter((_, i) => i !== idx)
      const lastAccess = { ...state.lastAccess }
      const char = state.characters[idx]
      if (char?.id) delete lastAccess[char.id]
      return {
        state: {
          ...state,
          characters,
          lastAccess,
          current: null
        }
      }
    }, false, 'deleteCharacter'),
  exportCharacters: () => {
    const { characters, lastAccess } = get().state
    const data = characters.map(c => ({ ...c, lastAccess: lastAccess[c.id] }))
    return JSON.stringify(data, null, 2)
  },
  importCharacters: data => {
    let parsed: unknown
    try {
      parsed = typeof data === 'string' ? JSON.parse(data) : data
    } catch {
      return false
    }
    if (!Array.isArray(parsed)) return false
    const sanitizedChars: Character[] = []
    const lastAccesses: Record<string, number> = {}
    parsed.forEach(c => {
      const id =
        typeof (c as { id?: string }).id === 'string'
          ? (c as { id: string }).id
          : crypto.randomUUID()
      const name = typeof (c as Character).name === 'string' ? (c as Character).name : ''
      const sheet =
        typeof (c as Character).sheet === 'object'
          ? { ...createSheet(), ...(c as Character).sheet }
          : createSheet()
      const inventory = Array.isArray((c as Character).inventory)
        ? (c as Character).inventory
        : []
      const scrolls = Array.isArray((c as Character).scrolls)
        ? (c as Character).scrolls
        : []
      const last = typeof (c as { lastAccess?: number }).lastAccess === 'number'
        ? (c as { lastAccess: number }).lastAccess
        : Date.now()
      sanitizedChars.push({ id, name, sheet, inventory, scrolls })
      lastAccesses[id] = last
    })
    set(
      ({ state }) => ({
        state: {
          ...state,
          characters: [...state.characters, ...sanitizedChars],
          lastAccess: { ...state.lastAccess, ...lastAccesses }
        }
      }),
      false,
      'importCharacters'
    )
    return true
  },
  roll: (notation: string, label = '') => {
    const roller = new DiceParser()
    const result = roller.roll(notation) as DiceRoll
    const total = result.total as number
    const entry: LogEntry = { label, notation, output: result.output, total }
    const diceResults: Array<{ type: DiceType; result: number }> = []
    const matches = [...notation.matchAll(/(\d*)d(4|6|8|12|20)/gi)]
    let matchIndex = 0
    for (const part of result.rolls as unknown[]) {
      if (part && typeof part === 'object' && 'rolls' in part) {
        const m = matches[matchIndex]
        const type = m ? (`d${m[2]}` as DiceType) : 'd20'
        for (const r of (part as { rolls: Array<{ value: number }> }).rolls) {
          diceResults.push({ type, result: r.value })
        }
        matchIndex++
      }
    }
    const { overlayTimeout } = get()
    if (overlayTimeout) clearTimeout(overlayTimeout)
    const timeout = setTimeout(() => {
      set(({ state }) => ({
        state: { ...state, overlay: { ...state.overlay, visible: false, roll: null } },
        overlayTimeout: null
      }), false, 'hideOverlay')
    }, 10000)
    set(({ state }) => ({
      state: {
        ...state,
        log: [entry, ...state.log],
        overlay: {
          message: result.output,
          roll: { dice: diceResults, total },
          visible: true
        }
      },
      overlayTimeout: timeout
    }), false, 'roll')
    return { total, output: result.output }
  },
  logInventory: message =>
    set(({ state }) => ({
      state: {
        ...state,
        log: [{ label: 'Inventory', output: message }, ...state.log]
      }
    }), false, 'logInventory'),
  applyEffect: effect => {
    const { target, value, subject, description } = effect
    const roller = new DiceParser()
    let amount = 0
    let output = value
    if (target !== 'condition') {
      try {
        const result = roller.roll(value) as DiceRoll
        amount = result.total as number
        output = result.output
      } catch {
        amount = Number(value) || 0
      }
    }
    const label = description || subject || target
    const message = `${label}: ${target === 'condition' ? value : output}`
    const { overlayTimeout } = get()
    if (overlayTimeout) clearTimeout(overlayTimeout)
    const timeout = setTimeout(() => {
      set(({ state }) => ({
        state: { ...state, overlay: { ...state.overlay, visible: false, roll: null } },
        overlayTimeout: null
      }), false, 'hideOverlay')
    }, 10000)
    set(({ state }) => {
      const newState = { ...state }
      if (target === 'hp') {
        newState.sheet = { ...newState.sheet, hp: newState.sheet.hp + amount }
      } else if (['str', 'agi', 'pre', 'tou'].includes(target)) {
        const key = target as 'str' | 'agi' | 'pre' | 'tou'
        newState.sheet = {
          ...newState.sheet,
          [key]: newState.sheet[key] + amount
        }
      } else if (target === 'item' && subject) {
        const items = [...newState.inventory]
        const idx = items.findIndex(i => i.name === subject)
        if (idx !== -1) {
          items[idx] = { ...items[idx], qty: items[idx].qty + amount }
        } else if (amount > 0) {
          items.push({ id: Date.now(), name: subject, qty: amount, notes: '' })
        }
        newState.inventory = items
      } else if (target === 'condition' && subject) {
        const conds = [...(newState.sheet.conditions || [])]
        if (value === 'remove' || value.startsWith('-')) {
          const idx = conds.indexOf(subject)
          if (idx !== -1) conds.splice(idx, 1)
        } else {
          if (!conds.includes(subject)) conds.push(subject)
        }
        newState.sheet = { ...newState.sheet, conditions: conds }
      }

      if (newState.current !== null) {
        const updated = [...newState.characters]
        let char = updated[newState.current]
        if (!char) {
          char = {
            id: crypto.randomUUID(),
            name: '',
            sheet: createSheet(),
            inventory: [],
            scrolls: []
          }
        } else if (!char.id) {
          char = { ...char, id: crypto.randomUUID() }
        }
        char = {
          ...char,
          sheet: newState.sheet,
          inventory: newState.inventory,
          scrolls: newState.scrolls
        }
        updated[newState.current] = char
        const lastAccess = { ...newState.lastAccess, [char.id]: Date.now() }
        newState.characters = updated
        newState.lastAccess = lastAccess
      }

      newState.log = [
        { label, notation: value, output, total: amount },
        ...newState.log
      ]
      newState.overlay = { message, roll: null, visible: true }
      return { state: newState, overlayTimeout: timeout }
    }, false, 'applyEffect')
    return amount
  },
  setDiceStyle: style =>
    set(({ state }) => ({ state: { ...state, diceStyle: style } }), false, 'setDiceStyle')
})

type PersistedState = {
  state: {
    characters: Character[]
    lastAccess: Record<string, number>
    log: LogEntry[]
    diceStyle: DiceStyle
  }
}

const storeWithPersist = persist(storeCreator, {
  name: 'openroll-store',
  storage: createJSONStorage<PersistedState>(() => localStorage),
  partialize: state => ({
    state: {
      characters: state.state.characters,
      lastAccess: state.state.lastAccess,
      log: state.state.log,
      diceStyle: state.state.diceStyle
    }
  }),
  merge: (persistedState, currentState) => {
    const merged = {
      ...currentState,
      state: {
        ...currentState.state,
        ...(persistedState as PersistedState).state
      }
    }
    const lastAccess = { ...merged.state.lastAccess }
    merged.state.characters = merged.state.characters.map(c => {
      const id = c.id || crypto.randomUUID()
      const char = c.id ? c : { ...c, id }
      if (!lastAccess[id]) lastAccess[id] = Date.now()
      return char
    })
    merged.state.lastAccess = lastAccess
    return merged
  }
})

const store = (import.meta.env.DEV
  ? devtools(storeWithPersist)
  : storeWithPersist) as StateCreator<
  GameContextValue,
  [],
  [["zustand/devtools", never], ["zustand/persist", unknown]]
>

export const useGameContext = create<GameContextValue>()(store)

// Alias for compatibility with previous naming
export type { GameContextValue as GameStore }

