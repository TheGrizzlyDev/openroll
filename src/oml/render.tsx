/* eslint react-refresh/only-export-components: off, react-hooks/rules-of-hooks: off */
import { useGameContext } from '../stores/GameContext'
import { parseOml, type OmlNode, type ApplyNode } from './parser'
import React from 'react'
import InventoryLookup from '../components/InventoryLookup'
import { Button } from '../components/ui'

export function renderNodes(
  nodes: OmlNode[],
  ctx: { roll: (_notation: string) => unknown; applyEffect: (_node: ApplyNode) => unknown }
) {
  return nodes.map((node, i) => {
    if (node.type === 'dice') {
      const label = node.description ?? node.notation
      return (
        <Button type="button" key={i} onClick={() => ctx.roll(node.notation)}>
          {label}
        </Button>
      )
    }
    if (node.type === 'apply') {
      const label = node.description ?? `${node.target} ${node.value}`
      return (
        <Button type="button" key={i} onClick={() => ctx.applyEffect(node)}>
          {label}
        </Button>
      )
    }
    if (node.type === 'if') {
      const { state } = useGameContext.getState()
      let active = -1
      for (let b = 0; b < node.branches.length; b++) {
        const br = node.branches[b]
        if (br.type === 'else') {
          if (active === -1) active = b
          break
        }
        if (br.predicate) {
          try {
            const fn = new Function('state', `with(state){with(state.sheet){return (${br.predicate});}}`)
            if (fn(state)) {
              active = b
              break
            }
          } catch {
            // ignore errors and treat as false
          }
        }
      }
      if (active === -1) active = node.branches.findIndex(br => br.type === 'else')
      return (
        <React.Fragment key={i}>
          {node.branches.map((br, idx) => (
            <span key={idx} style={idx === active ? undefined : { opacity: 0.5 }}>
              {renderNodes(br.children, ctx)}
            </span>
          ))}
        </React.Fragment>
      )
    }
    if (node.type === 'inventory') {
      return (
        <InventoryLookup
          key={i}
          description={node.description}
          attrs={node.attrs}
        />
      )
    }
    if (node.type === 'name') {
      const { state } = useGameContext.getState()
      return <React.Fragment key={i}>{state.sheet.name}</React.Fragment>
    }
    if (node.type === 'link') {
      const label = node.description ?? node.text
      return (
        <a key={i} href={node.url}>
          {label}
        </a>
      )
    }
    const parts = node.text.split(/(\r?\n)/)
    return (
      <span key={i} style={{ whiteSpace: 'break-spaces' }}>
        {parts.map((part, idx) =>
          /\r?\n/.test(part) ? <br key={idx} /> : <React.Fragment key={idx}>{part}</React.Fragment>
        )}
      </span>
    )
  })
}

export function renderOml(text: string, rollFn?: (_notation: string) => unknown) {
  const { roll, applyEffect } = useGameContext.getState()
  const doRoll = rollFn ?? roll
  const nodes = parseOml(text.trim())
  return <>{renderNodes(nodes, { roll: doRoll, applyEffect })}</>
}

export { parseOml, type OmlNode } from './parser'
