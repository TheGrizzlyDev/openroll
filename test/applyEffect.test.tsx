import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { describe, it, expect, beforeEach } from 'vitest'
import { renderOml } from '../src/oml/render'
import { useGameContext, type GameState } from '../src/GameContext'
import { createSheet } from '../src/morg_borg/sheet'

const setup = (state?: Partial<GameState>) => {
  const base: GameState = {
    characters: [],
    current: null,
    sheet: createSheet(),
    inventory: [],
    scrolls: [],
    log: [],
    activeTab: 'character',
    overlay: { message: '', visible: false }
  }
  useGameContext.setState({ state: { ...base, ...state }, overlayTimeout: null })
}

describe('apply effect handler', () => {
  beforeEach(() => setup())

  it('applies hp changes', () => {
    setup({ sheet: { ...createSheet(), hp: 10 } })
    const Test = () => <div>{renderOml('[apply "Fall" hp -1]')}</div>
    const { getByText } = render(<Test />)
    fireEvent.click(getByText('Fall'))
    const { state } = useGameContext.getState()
    expect(state.sheet.hp).toBe(9)
  })

  it('applies stat changes', () => {
    const Test = () => <div>{renderOml('[apply "Buff" str +2]')}</div>
    const { getByText } = render(<Test />)
    fireEvent.click(getByText('Buff'))
    const { state } = useGameContext.getState()
    expect(state.sheet.str).toBe(2)
  })

  it('applies item changes', () => {
    setup({ inventory: [{ id: 1, name: 'Potion', qty: 1, notes: '' }] })
    const Test = () => <div>{renderOml('[apply "Drink" item Potion -1]')}</div>
    const { getByText } = render(<Test />)
    fireEvent.click(getByText('Drink'))
    const { state } = useGameContext.getState()
    expect(state.inventory.find(i => i.name === 'Potion')?.qty).toBe(0)
  })

  it('applies condition changes', () => {
    const Test = () => (
      <div>{renderOml('[apply "Poison" condition poisoned add]')}</div>
    )
    const { getByText } = render(<Test />)
    fireEvent.click(getByText('Poison'))
    const { state } = useGameContext.getState()
    expect(state.sheet.conditions).toContain('poisoned')
  })
})
