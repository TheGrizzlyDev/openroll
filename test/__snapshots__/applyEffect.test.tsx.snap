// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`apply effect handler > renders correctly 1`] = `
<div>
  <div>
    <button
      class="bg-bg-alt border border-accent rounded-[var(--border-radius)] text-accent cursor-pointer font-body font-bold tracking-wider py-2 px-4 uppercase transition-colors hover:bg-accent hover:text-bg badge"
      type="button"
    >
      Buff
    </button>
  </div>
</div>
`;
